src/c++/perf_analyzer/periodic_concurrency_manager.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "periodic_concurrency_manager.h"
+
+namespace triton { namespace perfanalyzer {
+
+std::vector<RequestRecord>
+PeriodicConcurrencyManager::RunExperiment()
+{
+  AddConcurrentRequests(concurrency_range_.start);
+  WaitForRequestsToFinish();
+  return GetRequestRecords();
+}
+
+std::shared_ptr<IWorker>
+PeriodicConcurrencyManager::MakeWorker(
+    std::shared_ptr<ThreadStat> thread_stat,
+    std::shared_ptr<PeriodicConcurrencyWorker::ThreadConfig> thread_config)
+{
+  uint32_t id = workers_.size();
+
+  auto worker = std::make_shared<PeriodicConcurrencyWorker>(
+      id, thread_stat, thread_config, parser_, data_loader_, factory_,
+      on_sequence_model_, async_, max_concurrency_, using_json_data_,
+      streaming_, batch_size_, wake_signal_, wake_mutex_, active_threads_,
+      execute_, infer_data_manager_, sequence_manager_, request_period_,
+      period_completed_callback_, request_completed_callback_);
+  return worker;
+};
+
+void
+PeriodicConcurrencyManager::AddConcurrentRequests(
+    uint64_t num_concurrent_requests)
+{
+  for (size_t i = 0; i < num_concurrent_requests; i++) {
+    threads_stat_.emplace_back(std::make_shared<ThreadStat>());
```

Comment:
This would be more readable if you refactored the emplacements into a separate method

Reply:
Separated 👍